src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -183,6 +121,136 @@ public GeneratorDriver AddAdditionalTexts(ImmutableArray<AdditionalText> additio
+                        receiverCount++;
+                    }
+                    catch (Exception e)
+                    {
+                        generatorState = SetGeneratorException(MessageProvider, generatorState, generator, e, diagnosticsBag);
+                    }
+                }
+
+                stateBuilder.Add(generatorState);
+            }
+
+
+            // Run a syntax walk if any of the generators requested it
+            if (receiverCount > 0)
+            {
+                foreach (var tree in compilation.SyntaxTrees)
+                {
+                    var root = tree.GetRoot(cancellationToken);
+
+                    // https://github.com/dotnet/roslyn/issues/42629: should be possible to parallelize this
+                    for (int i = 0; i < walkerBuilder.Count; i++)
+                    {
+                        var walker = walkerBuilder[i];
+                        if (walker is object)
+                        {
+                            try
+                            {
+                                walker.Visit(root);
+                            }
+                            catch (Exception e)
+                            {
+                                stateBuilder[i] = SetGeneratorException(MessageProvider, stateBuilder[i], state.Generators[i], e, diagnosticsBag);
+                            }
+                        }
+                    }
+                }
+            }
+            walkerBuilder.Free();
+
+            // https://github.com/dotnet/roslyn/issues/42629: should be possible to parallelize this
+            for (int i = 0; i < state.Generators.Length; i++)
+            {
+                var generator = state.Generators[i];
+                var generatorState = stateBuilder[i];
+
+                // don't try and generate if initialization or syntax walk failed
+                if (!generatorState.Info.Initialized || generatorState.Exception is object)
+                {
+                    continue;
+                }
+
+                // we create a new context for each run of the generator. We'll never re-use existing state, only replace anything we have 
+                var context = new SourceGeneratorContext(compilation, state.AdditionalTexts.NullToEmpty(), state.OptionsProvider, generatorState.SyntaxReceiver);
+                try
+                {
+                    generator.Execute(context);
+                }
+                catch (Exception e)
+                {
+                    stateBuilder[i] = SetGeneratorException(MessageProvider, generatorState, generator, e, diagnosticsBag);
+                    continue;
+                }
+
+                (var sources, var diagnostics) = context.ToImmutableAndFree();
+                stateBuilder[i] = new GeneratorState(generatorState.Info, sources, ParseAdditionalSources(generator, sources, cancellationToken), diagnostics);
+                diagnosticsBag?.AddRange(diagnostics);
```

Comment:
>diagnosticsBag?.AddRange(diagnostics) [](start = 16, length = 37)

We're not including the diagnostic from `SetGeneratorException()` above. Is that expected? (It's not clear to me why diagnostics from exceptions are attached to the `GeneratorState` and some are simply added to the `diagnostics` collection.)

Reply:
We are, because `SetGeneratorException` sets it in the same bag. 

We always add diagnostics to the state, so you can retrieve them via `GetRunResults`. One of the `RunGenerator` variants also directly returns the diagnostics upfront, so if you pass in a bag we'll also collect them as we go and return them (rather than having to go pull them out of the state afterwards).



---
In reply to: [467297148](https://github.com/dotnet/roslyn/pull/46296#discussion_r467297148) [](ancestors = 467297148)